arch/arc/core/thread.c

```diff
@@ -253,3 +262,29 @@ int arch_float_enable(struct k_thread *thread, unsigned int options)
 	return 0;
 }
 #endif /* CONFIG_FPU && CONFIG_FPU_SHARING */
+
+#if !defined(CONFIG_MULTITHREADING)
+
+K_KERNEL_STACK_ARRAY_DECLARE(z_interrupt_stacks, CONFIG_MP_NUM_CPUS, CONFIG_ISR_STACK_SIZE);
+K_THREAD_STACK_DECLARE(z_main_stack, CONFIG_MAIN_STACK_SIZE);
+
+extern void z_main_no_multithreading_entry_wrapper(void *p1, void *p2, void *p3,
+						   void *main_stack, void *main_entry);
+
+FUNC_NORETURN void z_arc_switch_to_main_no_multithreading(k_thread_entry_t main_entry,
+							  void *p1, void *p2, void *p3)
+{
+	_kernel.cpus[0].id = 0;
+	_kernel.cpus[0].irq_stack = (Z_KERNEL_STACK_BUFFER(z_interrupt_stacks[0]) +
+				     K_KERNEL_STACK_SIZEOF(z_interrupt_stacks[0]));
```

Comment:
In case of CONFIG_MULTITHREADING is disabled we don't setup `_kernel.cpus[0].irq_stack` in generic code. That is fine as other architectures which support non-multithreading mode (well currently it's only ARM) don't use it. We initialize `_kernel.cpus[0].irq_stack`  to reuse the same code for selecting interrupt stack which we have for multithreading mode. 
Though there is a chance that I'll re-implement that differently in ARC interrupt code overhaul - so I don't think it worth to move it to generic code for now.